File: 프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.py
'''
0층으로 내려가는 데 필요한 마법의돌 최소 개수

풀이:
최대한 큰 수에서 나눠주는 것이 좋다.
1 - heapq? 시간초과
2 - 그리디? 낮은 자리부터 모듈러연산 하며 5보다 큰 경우면 더해주고, 아니면 뺴주는 식
    3-1. 만약 다음 자릿값이 5 ~ 9에 해당한다면 현재 자릿값을 10 에 도달하는 방향으로 마법의 돌을 사용한다.
    3-2. 만약 다음 자릿값이 0 ~ 4에 해당한다면 현재 자릿값을 0 에 도달하는 방향으로 마법의 돌을 사용한다.
'''

def solution(storey):
    answer = 0
    c = 10
    
    while True:
        v, m = storey // c, storey % c
        if m < 5:
            answer += m
        elif m > 5:
            v += 1
            answer += 10 - m
        else:
            tmp_m = v % c
            if tmp_m >= 5:
                v += 1
                answer += 10 - m
            else:
                answer += m
        if not v:
            break
        else:
            storey = v
        
    return answer

File: 프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.cpp
/*
버튼을 누르면 현재 층 + 버튼에 적힌 수 만큼 이동
0보다 작으면 움직이지 않음
0층으로 가기 위한 최소값을 리턴

5보다 작으면 빼는게 이득임
6이상이라면 더하는게 이득임
5라면, 다음 자리수를 확인해봐야함
*/


#include <string>
#include <vector>

using namespace std;

int solution(int storey) {
    int answer = 0;
    
    while (storey){
        int press = storey % 10;
        storey /= 10;
                
        if (!press)
            continue;
        if (press < 5){
            answer += press;
        }
        else if (press > 5){
            answer += 10 - press;
            storey += 1;
        }
        else {
            int tmp_press = storey % 10;
            if (tmp_press + 1 > 5){
                answer += 10 - press;
                storey += 1;
            }
            else {
                answer += press;
            }
        }
    }
    return answer;
}